### PR TITLE
Node 8.1.3

### LINKED ISSUE #302: UInt8Array check slowing things down
@addaleax the supported version of UInt8Array is slowing things down significantly here.
Can we implement the check in a faster way? 

cc https://bugs.chromium.org/p/v8/issues/detail?id=6519#c2

### LINKED ISSUE #305: TypeError undefined for chunk
I'm getting the following error:

```
/private/tmp/aurelia-orm/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:257
      if (typeof chunk !== 'string' && Object.getPrototypeOf(chunk) !== Buffer.prototype && !state.objectMode) {
                                              ^

TypeError: Cannot convert undefined or null to object
    at Function.getPrototypeOf (<anonymous>)
    at readableAddChunk (/private/tmp/aurelia-orm/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:257:47)
    at DestroyableTransform.Readable.push (/private/tmp/aurelia-orm/node_modules/through2/node_modules/readable-stream/lib/_stream_readable.js:237:10)
    at DestroyableTransform.Transform.push (/private/tmp/aurelia-orm/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:146:32)
    at /private/tmp/aurelia-orm/node_modules/aurelia-tools/src/build.js:43:12
    at action (/private/tmp/aurelia-orm/node_modules/breeze-dag/lib/dag.js:93:5)
    at Queue.process (/private/tmp/aurelia-orm/node_modules/breeze-queue/lib/queue.js:174:5)
    at next (/private/tmp/aurelia-orm/node_modules/breeze-queue/lib/queue.js:182:12)
    at done (/private/tmp/aurelia-orm/node_modules/breeze-dag/lib/dag.js:102:7)
    at /private/tmp/aurelia-orm/node_modules/aurelia-tools/src/build.js:44:7
    at action (/private/tmp/aurelia-orm/node_modules/breeze-dag/lib/dag.js:93:5)
    at Queue.process (/private/tmp/aurelia-orm/node_modules/breeze-queue/lib/queue.js:174:5)
    at module.exports (/private/tmp/aurelia-orm/node_modules/breeze-dag/lib/dag.js:118:9)
    at DestroyableTransform.endStream [as _flush] (/private/tmp/aurelia-orm/node_modules/aurelia-tools/src/build.js:42:5)
    at DestroyableTransform.<anonymous> (/private/tmp/aurelia-orm/node_modules/through2/node_modules/readable-stream/lib/_stream_transform.js:138:49)
    at Object.onceWrapper (events.js:293:19)
```

If I add the following line of code _before_ [the code on this line](https://github.com/nodejs/readable-stream/blob/master/lib/_stream_readable.js#L256), everything works as expected:

`if (!chunk) return`

Never mind the lack of respect for coding style, it was just to test.

I assume my addition isn't needed, so I'm hoping you can tell me what's going on here. I have no idea.

### PR DESCRIPTION
**Pull request by mcollina**
Includes #304 and #303.

Fixes #302 
Fixes #305

### CODE REVIEW

File: build/files.js [hunk 1]
```diff
@@ -180,11 +180,12 @@ const headRegexp = /(^module.exports = \w+;?)/m
       /(?:var|const) Buffer = require\('buffer'\)\.Buffer;/
     , `/*<replacement>*/
   const Buffer = require('safe-buffer').Buffer
+  const OurUint8Array = global.Uint8Array || function () {}
 function _uint8ArrayToBuffer(chunk) {
    return Buffer.from(chunk);
 }
 function _isUint8Array(obj) {
-  return Object.prototype.toString.call(obj) === '[object Uint8Array]' || Buffer.isBuffer(obj);
+  return Buffer.isBuffer(obj) || obj instanceof OurUint8Array;
```
Comment (yoshuawuyts) [inline build/files.js]: Just to check: this is meant to always pass in environments that don't have `global.Uint8Array`, right?

Comment (yoshuawuyts) [inline build/files.js]: e.g. resolves to `obj instanceOf Function`

File: lib/_stream_writable.js [hunk 2]
```diff
@@ -660,4 +661,4 @@ Writable.prototype._undestroy = destroyImpl.undestroy;
 Writable.prototype._destroy = function (err, cb) {
   this.end();
   cb(err);
-};
+};
```
Comment (yoshuawuyts) [inline lib/_stream_writable.js]: lil nit

File: test/parallel/test-stream-objectmode-undefined.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      _stream2.push(chunk);
+    })
+  });
+
+  _stream2.on('data', common.mustCall(function (chunk) {
+    assert.strictEqual(chunk, undefined);
+  }));
+
+  _stream2.write(undefined);
+}
```
Comment (yoshuawuyts) [inline test/parallel/test-stream-objectmode-undefined.js]: lil nit

File: build/files.js (hunk 1 above)
Comment (mcollina) [inline build/files.js]: Yes exactly.

**Comment by mcollina** [comment]
The nits are out of babel, so there is not much we can do about them

**Pull request merged**